src/basilisp/lang/tagged.py

```diff
@@ -0,0 +1,85 @@
+from typing import (
+    Optional,
+    TypeVar,
+)
+
+from typing_extensions import Unpack
+
+from basilisp.lang.interfaces import (
+    ILispObject,
+    ILookup,
+)
+
+from basilisp.lang.keyword import Keyword
+from basilisp.lang.obj import PrintSettings, lrepr
+from basilisp.lang.symbol import Symbol
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+class TaggedLiteral(
+    ILispObject,
+    ILookup[K, V],
+):
+    """Basilisp TaggedLiteral. https://clojure.org/reference/reader#tagged_literals
+    """
+
+    __slots__ = ("_tag", "_form", "_hash")
+
+    def __init__(
+        self, tag: Symbol, form
+    ) -> None:
+        self._tag = tag
+        self._form = form
+        self._hash = -1
+
+    @property
+    def tag(self) -> Symbol:
+        return self._tag
+
+    @property
+    def form(self):
+        return self._form
+
+    def __bool__(self):
+        return True
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, TaggedLiteral):
+            return NotImplemented
+        return self._tag == other._tag and self._form == other._form
+
+    def __hash__(self):
+        if self._hash == -1:
+            self._hash = hash((self._tag, self._form))
+        return self._hash
+
+    def __getitem__(self, item):
+        if item == Keyword("tag"):
```

Comment:
I would probably just create `:tag` and `:form` as keywords using `basilisp.lang.keyword.keyword` function (which performs the interning) and then store them in this module as private constants like:

```python
_TAG_KW = kw.keyword("tag")
_FORM_KW = kw.keyword("form")
```

Then you can reference the "constants" from here.